drivers: serial: stm32: use PM constraints to prevent suspension

gmarull opened the pull request:
In the current implementation, the STM32 UART driver required to enable
`CONFIG_PM_DEVICE` when `CONFIG_PM=y` to function properly. The main
reason is that in some situations, like in polling mode, transmissions
are not fully synchronous. That is, a byte is pushed to the _queue_ if
it is empty, and then the function returns without waiting for it to be
transmitted to the wire. This makes sense to make things like per-byte
transmission efficient. However, this introduces a problem: the system
may reach idle state, and so enter low power modes before the UART has
actually finished processing the last data in the queue. If this happens,
communications can be interrupted, or garbage data may be put into the
UART line.

The proposed solution in this patch uses PM constraints to solve this
problem. For the IRQ/DMA case it is easy since we can set the constraint
before transmission start, and when the completion (TC) interrupt is
received we can clear it. However, the polling mode did not have the
capability to signal the completion. For this case, a simpler IRQ
routine is provided to just release the constraint. As a result, the PM
hooks are not required, and so the system can operate with just `CONFIG_PM`.

Fixes #38382 
Fixes #38437

@erwango @FRASTM seems to fix the issue on my side, but I may have missed things. `CONFIG_PM_DEVICE=y` default for STM32 should likely be reviewed.

gmarull committed:
drivers: serial: stm32: use PM constraints to prevent suspension

In the current implementation the STM32 UART driver required to enable
`CONFIG_PM_DEVICE` when `CONFIG_PM=y` to function properly. The main
reason is that in some situations, like in polling mode, transmissions
are not fully synchronous. That is, a byte is pushed to the _queue_ if
it is empty and then the function returns without waiting for it to be
transmitted to the wire. This makes sense to make things like per-byte
transmission efficient. However, this introduces a problem: the system
may reach idle state, and so enter low power modes before the UART has
actually finished the last data in the queue. If this happens,
communications can be interrupted or garbage data may be put into the
UART line.

The proposed solution in this patch uses PM constraints to solve this
problem. For the IRQ/DMA case it is easy since we can set the constraint
before transmission start, and when the completion (TC) interrupt is
received we can clear it. However, the polling mode did not have the
capability to signal the completion. For this case, a simpler IRQ
routine is provided to just release the constraint. As a result, the PM
hooks are not required and so system can operate with just `CONFIG_PM`.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

ceolin commented:
Looks good removing the pm control callback since it was not doing PM and move that logic to the places where they make sense. One question, isn't possible to use `pm_device_busy_set` ? This way the SoC will still go sleep but this device will remain active, if not using constraint is the right thing to do.

gmarull commented:
> Looks good removing the pm control callback since it was not doing PM and move that logic to the places where they make sense. One question, isn't possible to use `pm_device_busy_set` ? This way the SoC will still go sleep but this device will remain active, if not using constraint is the right thing to do.

The problem is that in low power modes the UART is not operational as far as I understand:

![image](https://user-images.githubusercontent.com/25011557/132765745-4046ded8-ce0b-4200-a671-e34eafb28911.jpeg)

erwango reviewed (changes requested):
Thanks for that proposal, that's probably the best compromise given previous discussions.
I'm fine with the change if @dcpleung agrees as well (specially to have IRQ logic in poll_out case).
Only request would be to set the full logic under CONFIG_PM.

FRASTM commented:
I can confirm, it Fixes the https://github.com/zephyrproject-rtos/zephyr/issues/38382

erwango commented on drivers/serial/uart_stm32.c:
```diff
@@ -451,12 +452,13 @@ static void uart_stm32_poll_out(const struct device *dev,
 	USART_TypeDef *UartInstance = UART_STRUCT(dev);
 
+	/* do not allow system to suspend until transmission has completed */
+	pm_constraint_set(PM_STATE_SUSPEND_TO_IDLE);
```
@ceolin, isn't it possible to have empty `pm_constraint_` definitions when PM is not defined?
This would avoid cluttering the code for such cases

ceolin replied:
https://github.com/zephyrproject-rtos/zephyr/pull/38472

gmarull reviewed (commented):
@erwango There's still one path that I think needs to be handled: If user disables TX irq (uart_stm32_irq_tx_disable), the constraint will never be released. Should that function just toggle a flag (to avoid callback being fired)?

erwango commented:
> uart_stm32_irq_tx_disable

Good catch.  So, the use case would be that user would use both `fifo_fill` and `poll_out` on the same device and for some reason (concurrent threads for instance), uart_irq_tx_disable happens during `poll_out` transaction. Seems plausible indeed.

According to API, using a flag is not sufficient " * @brief Disable TX interrupt in IER.".
Though, my feeling is that API is going too much deep into driver implementation, so it should be fine.
I guess that last word on this would be to @dcpleung .

erwango commented on drivers/serial/uart_stm32.c:
```diff
@@ -455,7 +456,12 @@ static void uart_stm32_poll_out(const struct device *dev,
+#ifdef CONFIG_PM
+	/* do not allow system to suspend until transmission has completed */
+	pm_constraint_set(PM_STATE_SUSPEND_TO_IDLE);
+#else
 	LL_USART_ClearFlag_TC(UartInstance);
+#endif /* CONFIG_PM */
```
Minor enhancement proposal:
According to RM, use of TC is "to avoid corrupting the last transmission when the USART is
disabled or enters Halt mode.".
Since we're not checking it when PM=n, I think it should be safe removing it completely from here.

gmarull replied:
done

gmarull committed:
drivers: serial: stm32: remove unnecessary flag clear

According to reference manual, use of TC is "to avoid corrupting the
last transmission when the USART is disabled or enters Halt mode.". It
is safe to remove it since it is not checked when CONFIG_PM=n.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

FRASTM commented on drivers/serial/uart_stm32.c:
```diff
@@ -1499,10 +1514,15 @@ static void uart_stm32_irq_config_func_##index(const struct device *dev)	\
+#elif defined(CONFIG_PM)
 #define STM32_UART_IRQ_HANDLER_FUNC(index)
-#define STM32_UART_IRQ_HANDLER(index)
+#define STM32_UART_POLL_IRQ_HANDLER_FUNC(index)				\
```
in case the soc (like stm32l476) has no CONFIG_UART_INTERRUPT_DRIVEN, no CONFIG_UART_ASYNC_API, no CONFIG_PM, then the build is raising an error:
`drivers/serial/uart_stm32.c:1547:3: error: implicit declaration of function 'STM32_UART_IRQ_HANDLER_FUNC'; did you mean 'STM32_UART_IRQ_HANDLER_DECL'? [-Werror=implicit-function-declaration]`
--> for example:  _west build -p auto -b nucleo_l476rg `tests/kernel/timer/timer_api/_

gmarull replied:
good catch, should be fixed

dcpleung reviewed (approved):
LGTM. Do you envision a use case where interrupt is totally disabled (e.g. interrupt controller not begin setup at all)? The closest I could think of is a bootloader, but PM is not needed for that.

gmarull commented:
> LGTM. Do you envision a use case where interrupt is totally disabled (e.g. interrupt controller not begin setup at all)? The closest I could think of is a bootloader, but PM is not needed for that.

A user can call `uart_irq_tx_disable`, however, when PM is enabled we always need to handle TX complete interrupt (to release PM constraint). I assume that when users call `uart_irq_tx_disable` they just expect that the user callback is not fired, am I right?

dcpleung commented:
> A user can call `uart_irq_tx_disable`, however, when PM is enabled we always need to handle TX complete interrupt (to release PM constraint). I assume that when users call `uart_irq_tx_disable` they just expect that the user callback is not fired, am I right?

Some might actually expects that interrupts to be totally disabled.

Though, what I was thinking was that whether there are any use cases where interrupts are not configured at all at build time. So at runtime, the interrupt controller is not being setup and interrupts not being routed correctly. In this case, you won't get any interrupts at all. However, this is a rather unique use case.

erwango commented:
> > A user can call `uart_irq_tx_disable`, however, when PM is enabled we always need to handle TX complete interrupt (to release PM constraint). I assume that when users call `uart_irq_tx_disable` they just expect that the user callback is not fired, am I right?
> 
> Some might actually expects that interrupts to be totally disabled.
> 
> Though, what I was thinking was that whether there are any use cases where interrupts are not configured at all at build time. So at runtime, the interrupt controller is not being setup and interrupts not being routed correctly. In this case, you won't get any interrupts at all. However, this is a rather unique use case.

In case PM is disabled, the poll operations will work w disabled IRQ. I guess this should be sufficient for these specifci cases, right?

erwango commented on drivers/serial/uart_stm32.c:
```diff
@@ -455,7 +456,10 @@ static void uart_stm32_poll_out(const struct device *dev,
 	}
 
-	LL_USART_ClearFlag_TC(UartInstance);
+#ifdef CONFIG_PM
```
We should be able to remove these, thanks to https://github.com/zephyrproject-rtos/zephyr/pull/38472

erwango commented on drivers/serial/uart_stm32.c:
```diff
@@ -792,16 +803,24 @@ static void uart_stm32_dma_rx_flush(const struct device *dev)
+	if (LL_USART_IsActiveFlag_TC(UartInstance)) {
+		LL_USART_ClearFlag_TC(UartInstance);
+
+#ifdef CONFIG_PM
```
ditto

erwango commented on drivers/serial/uart_stm32.c:
```diff
@@ -823,13 +842,29 @@ static void uart_stm32_isr(const struct device *dev)
 		/* Generate TX_DONE event when transmission is done */
 		async_evt_tx_done(data);
+
+#ifdef CONFIG_PM
```
ditto

erwango commented on drivers/serial/uart_stm32.c:
```diff
@@ -1052,6 +1087,11 @@ static int uart_stm32_async_tx(const struct device *dev,
 	/* Start TX timer */
 	async_timer_start(&data->dma_tx.timeout_work, data->dma_tx.timeout);
 
+#ifdef CONFIG_PM
```
ditto

erwango commented on drivers/serial/uart_stm32.c:
```diff
@@ -1501,10 +1516,22 @@ static void uart_stm32_irq_config_func_##index(const struct device *dev)	\
 #define STM32_UART_IRQ_HANDLER(index)
 #endif
 
+#if defined(CONFIG_UART_INTERRUPT_DRIVEN) || defined(CONFIG_UART_ASYNC_API)
+#define STM32_UART_IRQ_HANDLER_FUNC(index)				\
+	.irq_config_func = uart_stm32_irq_config_func_##index,
+#define STM32_UART_POLL_IRQ_HANDLER_FUNC(index)
+#elif defined(CONFIG_PM)
+#define STM32_UART_IRQ_HANDLER_FUNC(index)
+#define STM32_UART_POLL_IRQ_HANDLER_FUNC(index)				\
+	.irq_config_func = uart_stm32_irq_config_func_##index,
+#else
+#define STM32_UART_IRQ_HANDLER_FUNC(index)
+#define STM32_UART_POLL_IRQ_HANDLER_FUNC(index)
+#endif
```
Nitpick (I'm thinking to the ones that will read these lines for the first time)

```suggestion
#if defined(CONFIG_UART_INTERRUPT_DRIVEN) || defined(CONFIG_UART_ASYNC_API)
#define STM32_UART_IRQ_HANDLER_FUNC(index)				\
	.irq_config_func = uart_stm32_irq_config_func_##index,
#define STM32_UART_POLL_IRQ_HANDLER_FUNC(index) /* Not used */
#elif defined(CONFIG_PM)
#define STM32_UART_IRQ_HANDLER_FUNC(index)     /* Not used */
#define STM32_UART_POLL_IRQ_HANDLER_FUNC(index)			\
	.irq_config_func = uart_stm32_irq_config_func_##index,
#else
#define STM32_UART_IRQ_HANDLER_FUNC(index)      /* Not used */
#define STM32_UART_POLL_IRQ_HANDLER_FUNC(index) /* Not used */
#endif
```

gmarull replied:
done

gmarull commented:
> > > A user can call `uart_irq_tx_disable`, however, when PM is enabled we always need to handle TX complete interrupt (to release PM constraint). I assume that when users call `uart_irq_tx_disable` they just expect that the user callback is not fired, am I right?
> > 
> > 
> > Some might actually expects that interrupts to be totally disabled.
> > Though, what I was thinking was that whether there are any use cases where interrupts are not configured at all at build time. So at runtime, the interrupt controller is not being setup and interrupts not being routed correctly. In this case, you won't get any interrupts at all. However, this is a rather unique use case.
> 
> In case PM is disabled, the poll operations will work w disabled IRQ. I guess this should be sufficient for these specifci cases, right?

The problem only happens if `uart_irq_tx_disable` is called. If `uart_stm32_fifo_fill` is called afterwards, or polling API is used, then PM constraint will not be released. I'm not sure if this path makes sense in practice, though.

erwango reviewed (approved):
LGTM on my side, thanks!
 I let you decide when ready to merge and remove the DNM flag.
